txnkv/transaction/pessimistic.go

```diff
@@ -147,152 +165,384 @@ func (action actionPessimisticLock) handleSingleBatch(c *twoPhaseCommitter, bo *
+		lock, err1 := txnlock.ExtractLockFromKeyErr(keyErr)
+		if err1 != nil {
+			return true, err1
+		}
+		locks = append(locks, lock)
+		if !keyErr.GetLocked().GetSkipResolvingLock() {
+			skipResolvingLocks = false
+		}
+	}
+
+	if skipResolvingLocks {
+		// Retry immediately.
+		return false, nil
+	}
+
+	// Because we already waited on tikv, no need to Backoff here.
+	// tikv default will wait 3s(also the maximum wait value) when lock error occurs
+	if diagCtx.resolvingRecordToken == nil {
+		token := c.store.GetLockResolver().RecordResolvingLocks(locks, c.startTS)
+		diagCtx.resolvingRecordToken = &token
+	} else {
+		c.store.GetLockResolver().UpdateResolvingLocks(locks, c.startTS, *diagCtx.resolvingRecordToken)
+	}
+	resolveLockOpts := txnlock.ResolveLocksOptions{
+		CallerStartTS: 0,
+		Locks:         locks,
+		Detail:        &action.LockCtx.Stats.ResolveLock,
+	}
+	resolveLockRes, err := c.store.GetLockResolver().ResolveLocksWithOpts(bo, resolveLockOpts)
+	if err != nil {
+		return true, err
+	}
+
+	// If msBeforeTxnExpired is not zero, it means there are still locks blocking us acquiring
+	// the pessimistic lock. We should return acquire fail with nowait set or timeout error if necessary.
+	if resolveLockRes.TTL > 0 {
+		if action.LockWaitTime() == kv.LockNoWait {
+			return true, errors.WithStack(tikverr.ErrLockAcquireFailAndNoWaitSet)
+		} else if action.LockWaitTime() == kv.LockAlwaysWait {
+			// do nothing but keep wait
+		} else {
+			// the lockWaitTime is set, we should return wait timeout if we are still blocked by a lock
+			if time.Since(action.WaitStartTime).Milliseconds() >= action.LockWaitTime() {
+				return true, errors.WithStack(tikverr.ErrLockWaitTimeout)
 			}
+		}
+		if action.LockCtx.PessimisticLockWaited != nil {
+			atomic.StoreInt32(action.LockCtx.PessimisticLockWaited, 1)
+		}
+	}
+
+	return false, nil
+}
 
-			// Extract lock from key error
-			lock, err1 := txnlock.ExtractLockFromKeyErr(keyErr)
-			if err1 != nil {
-				return err1
+func (action actionPessimisticLock) handlePessimisticLockResponseLockFirstMode(c *twoPhaseCommitter, bo *retry.Backoffer, batch *batchMutations, mutationsPb []*kvrpcpb.Mutation, resp *tikvrpc.Response, diagCtx *diagnosticContext) (finished bool, retryMutations CommitterMutations, err error) {
+	regionErr, err := resp.GetRegionError()
+	if err != nil {
+		return true, nil, err
+	}
+	if resp.Resp == nil {
+		return true, nil, errors.WithStack(tikverr.ErrBodyMissing)
+	}
+	lockResp := resp.Resp.(*kvrpcpb.PessimisticLockResponse)
+	failedMutations := NewPlainMutations(0)
+	keyErrs := lockResp.GetErrors()
+
+	if batch.isPrimary && len(lockResp.Results) > 0 && lockResp.Results[batch.primaryIndex].Type != kvrpcpb.PessimisticLockKeyResultType_Failed {
+		// After locking the primary key, we should protect the primary lock from expiring
+		// now in case locking the remaining keys take a long time.
+		c.run(c, action.LockCtx)
+	}
+
+	for i, res := range lockResp.Results {
```

Comment:
As we've limited the use case to be a single key lock, seems these extra checks could be ignored by now?

Reply:
Yes.